packages/mui-material/src/OverridableComponentAugmentation.ts

```diff
@@ -0,0 +1,44 @@
+import * as React from 'react';
+import { DistributiveOmit } from '@mui/types';
+
+declare module '@mui/material/OverridableComponent' {
+  /**
+   * A component whose root component can be controlled via a `component` prop.
+   *
+   * Adjusts valid props based on the type of `component`.
+   */
+  interface OverridableComponent<M extends OverridableTypeMap> {
+    <C extends React.ElementType>(
+      props: {
+        /**
+         * The component used for the root node.
+         * Either a string to use a HTML element or a component.
+         */
+        component: C;
+      } & OverridePropsVer2<M, C>,
+    ): JSX.Element;
+    (props: DefaultComponentPropsVer2<M>): JSX.Element;
+  }
+
+  /**
+   * Props of the component if `component={Component}` is used.
+   */
+  // prettier-ignore
+  type OverridePropsVer2<
+    M extends OverridableTypeMap,
+    C extends React.ElementType,
+  > = (
+    & BaseProps<M>
+    & DistributiveOmit<React.ComponentPropsWithoutRef<C>, keyof BaseProps<M>>
+    & { ref?: React.Ref<Element> }
```

Comment:
Would this be better?
```suggestion
    & { ref?: C extends keyof JSX.IntrinsicElements ? React.Ref<C> : React.Ref<Element> }
```

Reply:
Other than this, looks good to me.

Reply:
Let me compare the time it takes to do the type checks and will report back :)

Reply:
Actually these is a big difference:

![image](https://user-images.githubusercontent.com/4512430/187184793-a2bb902c-ae5d-429b-b658-6b1418f5a463.png) ~ 100ms for checking App.tsx

vs

![image](https://user-images.githubusercontent.com/4512430/187184918-6e7f6676-3532-494e-bc88-ff2f32eaff5e.png) ~ 1344md for checking App.tsx

I would ship the initial version without this check.